README
This is Perl module Text::Xslate.

NAME

Text::Xslate - Scalable template engine for Perl5

SYNOPSIS

    use Text::Xslate;

    my $tx = Text::Xslate->new();

    my %vars = (
        title => 'A list of books',
        books => [
            { title => 'Islands in the stream' },
            { title => 'Programming Perl'      },
            # ...
        ],
    );

    my $template = q{
    <h1><: $title :></h1>
    <ul>
    : for $books -> $book {
        <li><: $book.title :></li>
    : } # for
    </ul>
    };

    print $tx->render_string($template, \%vars);

INSTALLATION

Install cpanm (App::cpanminus) and then run the following command to install
Xslate:

    $ cpanm Text::Xslate

If you get the distribution, unpack it and build it as per the usual:

    $ tar xzf Text-Xslate-{version}.tar.gz
    $ cd Text-Xslate-{version}
    $ perl Makefile.PL
    $ make && make test

Then install it:

    $ make install

If you want to install it from the repository, you must install authoring
tools.

    $ cpanm < author/requires.cpanm

DOCUMENTATION

Text::Xslate documentation is available as in POD. So you can do:

    $ perldoc Text::Xslate

to read the documentation online with your favorite pager.

RESOURCE

    web site:     http://xslate.org/
    repositories: http://github.com/xslate
    mailing list: http://groups.google.com/group/xslate
    irc         : irc://irc.perl.org/#xslate

LICENSE AND COPYRIGHT

Copyright (c) 2010, Fuji, Goro (gfx). All rights reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.



Changes
Revision history for Perl extension Text::Xslate

{{$NEXT}}

v3.5.9 2021-10-31T01:10:44Z
    - Fix t/010_internals/008_files.t (#206)

v3.5.8 2020-06-01T15:17:51Z
    - Fix "Argument list too long" in test (greeneg #202, #203, #204)

v3.5.7 2019-11-25T16:15:32Z
    - Fix a segfault on array join (zanyou #199, #200)

v3.5.6 2018-01-08T06:39:38Z
    - Fix build on perl 5.8.8 (wyoung #192, #193)

v3.5.5 2018-01-05T02:53:17Z
    - Ensure that test cleans up after itself (jkeenan #190)

v3.5.4 2017-12-19T14:06:41Z
    - Guarantee that test cleans up after itself (jkeenan #189)

v3.5.3 2017-12-03T05:18:51Z
    - Fix tests for windows again (#188)

v3.5.2 2017-12-02T08:19:33Z
    - Fix tests for windows (zdm #186, #187)

v3.5.1 2017-11-30T16:09:33Z
    - Fix packaging issue; explicitly declare version for Text::Xslate::Engine

v3.5.0 2017-11-30T15:57:50Z
    - Migrate to using minil (#179)
    - Improve perl 5.8 support (#182)
    - Use version->declare() to declare $VERSION (#182)
    - Make `prove -br t` work with dot-not-in-INC perls (#184)
    - Run tests against both XS and pureperl (#185)
    - Fix typo (anirvan #181)

3.4.0 2017-01-20 15:00:00+0900
    - Fix tied hash issue(#173)
    - Fix encoding issue(#160)
    - Fix segmentation fault issue by accessing uninitialized array element(#159)

3.3.9 2015-12-18 22:15:00+0900
    - Fix test for Windows
      3.3.8 change breaks Windows

3.3.8 2015-12-18 16:20:00+0900
    - Fix test for DragonflyBSD
    - Update document

3.3.7 2015-08-28 13:45:00+0900
    - Fix for older Perl 5.8.8 or lower(#145)
    - Enable 5.8 tests again

3.3.6 2015-08-25 13:50:00+0900
    - Fix issue 'include' makes stack pointer incorrect(#130)

3.3.5 2015-08-05 18:50:00+0900
    - Update Mouse version for Perl 5.22 or higher

3.3.4 2015-03-24 23:21:57+0900
    - Fix typos in document
    - Introduce $Text::Xslate::DEFAULT_CACHE_DIR

3.3.3 2014-08-04 15:50:00+0900
    - No code changes from 3.3.2, just re-packaging.

3.3.2 2014-08-04 12:40:00+0900
   [IMPROVEMENT]
    - #127 Improve generating temporary file name

3.3.1 2014-08-03 11:45:40+0900
    [ANNOUNCE]
    - Now Xslate has new co-maintainers (those who can maintain and release
      Xslate): lestrrat, syohex, and tokuhirom

    [BUG FIXES]
    - #126 Fix a typo in naming tempfiles (Hugmeir)

    [DOCMENT]
    - #125 Update document on Text::Xslate->bridge() and Builtin.pod
      (Mike Raynham)

3.3.0 2014-08-03 11:54:02+0900
    (mis-packaging release)

3.2.5 2014-07-15 08:18:31+0900
    [TEST FIXES]
        - No CGI.pm, which is no longer a standard module

3.2.4 2014-04-27 17:18:15+0900
    [TEST FIXES]
        - Fix #122, #120 testing issue on windows (syohex)
    [FEATURES]
        - Add $array.first() and $array.last() methods (#116, #118 / shyohex)

3.2.3 2014-04-23 07:32:39+0900
    - Made a mistake in the release engineering, re-packaged on the correct
      status.

3.2.2 2014-04-23 07:29:42+0900
    [BUG FIXES]
    - Fix #105 SEGV on blead (reported by tokuhirom, fixed by syohex in #117)

3.2.1 2014-04-17 07:55:38+0900
    [BUG FIXES]
    - Fix #111 (by syohex in #113) inputting "0" made a wrong result

3.2.0 2014-04-04 07:39:59+0900
    [BUG FIXES]
    - Fix #107, #109 HashWithDefault should use "exists" (yappo, tokuhirom)

3.1.2 2014-02-20 21:09:47+0900
    [TEST]
    - Add a test for github issue #105, which will fail on Perl 5.19.x

3.1.1 2014-01-24 07:50:52+0900
    [DOCUMENTS]
    - Fix some typos (#102)
    - Add an explanation of `validate()` (#101)

    [BUG FIXES]
    - Fix a race condition on making cache dirs (#103)

3.1.0 2013-11-16 16:46:35+0900
    [BUG FIXES]
    - Close #96; $/ affected the parse() method
    
    [FEATURES]
    - Add $xslate->validate($file) method to check template syntax

3.0.2 2013-11-15 21:56:53+0900
    [BUG FIXES]
    - Fix a mojibake issue where utf8::upgrade() was always called when
      loading caches (hanabukuro++)

3.0.1 2013-11-04 12:27:51+0900
    [TEST FIXES]
    - Fix a test that might fail on a slow machine like Raspberry Pi
      (Getty++)

3.0.0 2013-10-18 08:59:22+0900
    - No code changes from 2.1.0, just re-packaging for package managers

2.1.0 2013-10-17 22:18:21+0900
    [BUG FIXES]
    - Fix an issue that multi-bytes string literals used for a hash key
      was not dealt as a text string (@Niratama++)
      WARNING: this change could break your code if you use multi-byte
               text string as a hash key.

2.0010 2013-10-07 21:56:12+0900
    [BUG FIXES]
    - Fix an issue that vpath with text-strings raised errors
      on newer perls due to specification changes in PerlIO::scalar (#90)
    - Documentation tweaks (#84, #86)

2.0009 2013-07-08 10:49:47-0700
    [BUG FIXES]
    - Fix a tied hash issue on string concat, thanks to zxchris (#81)

2.0008 2013-07-06 00:09:43-0700
    [BUG FIXES]
    - Fix a chache issue that UTF8 flags were not turned on
      as expected (#82)
    - Fix Metakolon doc (#83)

2.0007 2013-06-21 13:16:16-0700
    [DOCUMENTS]
    - Add doc about chomping newlines (<:- ... -:>)

    [NEW FEATURES]
    - Add pre_process_handler to pre-process template content (bluescreen)

2.0006 2013-06-02 22:58:27
    [BUG FIXES]
    - Fix an issue that the state of VM broke on exception
      reported by nihen

2.0005 2013-06-02 12:23:15
    [BUG FIXES]
    - Fix a Multiplexed WRAPPER issue reportedy by @kawamuray (#79)
      Those who use WRAPPER, template cascading, or anonymous functions
      (lambdas) are strongly recommended to upgrade Xslate.

2.0004 2013-05-29 20:27:07
    [TEST FIXES]
    - Template string must be bytes (#78)

2.0003 2013-05-10 12:59:11
    [BUG FIXES]
    - Fix a crash on compiling templates (#77)

2.0002 2013-04-26 12:51:47
    [BUG FIXES]
    - Support perl-blead (5.17.11)

2.0001 2013-04-24 16:24:22
    [FEATURES]
    - Support PUREPERL_ONLY build
      See the Lancaster Consensu:

xslate_ops.h
/* This file is automatically generated by tool/opcode.pl.
 * ANY CHANGES WILL BE LOST!
 */

/* forward decl for Xslate opcodes */
TXC(noop);
TXC(move_to_sb);
TXC(move_from_sb);
TXC_w_var(save_to_lvar);
TXC_w_var(load_lvar_to_sb);
TXC_w_key(local_s);
TXC(push);
TXC(pushmark);
TXC(nil);
TXC_w_sv(literal);
TXC_w_int(literal_i);
TXC_w_key(fetch_s); /* fetch a field from the top */
TXC_w_var(fetch_lvar);
TXC(fetch_field); /* fetch a field from a variable (bin operator) */
TXC_w_key(fetch_field_s); /* fetch a field from a variable (for literal) */
TXC(print);
TXC(print_raw);
TXC_w_sv(print_raw_s);
TXC(include);
TXC_w_var(for_start);
TXC_goto(for_iter);
TXC(add);
TXC(sub);
TXC(mul);
TXC(div);
TXC(mod);
TXC_w_sv(concat);
TXC_goto(and);
TXC_goto(dand);
TXC_goto(or);
TXC_goto(dor);
TXC(not);
TXC(plus); /* unary plus */
TXC(minus); /* unary minus */
TXC(eq);
TXC(ne);
TXC(lt);
TXC(le);
TXC(gt);
TXC(ge);
TXC(macrocall);
TXC_w_key(macro_begin);
TXC(macro_end);
TXC_w_key(macro);
TXC_w_key(function);
TXC(funcall);
TXC_w_key(methodcall_s);
TXC(make_array);
TXC(make_hash);
TXC(enter);
TXC(leave);
TXC_goto(goto);
TXC_w_sv(depend); /* tell the vm to dependent template files */
TXC(end);

enum tx_opcode_t {
    TXOP_noop, /* 0 */
    TXOP_move_to_sb, /* 1 */
    TXOP_move_from_sb, /* 2 */
    TXOP_save_to_lvar, /* 3 */
    TXOP_load_lvar_to_sb, /* 4 */
    TXOP_local_s, /* 5 */
    TXOP_push, /* 6 */
    TXOP_pushmark, /* 7 */
    TXOP_nil, /* 8 */
    TXOP_literal, /* 9 */
    TXOP_literal_i, /* 10 */
    TXOP_fetch_s, /* 11 */
    TXOP_fetch_lvar, /* 12 */
    TXOP_fetch_field, /* 13 */
    TXOP_fetch_field_s, /* 14 */
    TXOP_print, /* 15 */
    TXOP_print_raw, /* 16 */
    TXOP_print_raw_s, /* 17 */
    TXOP_include, /* 18 */
    TXOP_for_start, /* 19 */
    TXOP_for_iter, /* 20 */
    TXOP_add, /* 21 */
    TXOP_sub, /* 22 */
    TXOP_mul, /* 23 */
    TXOP_div, /* 24 */
    TXOP_mod, /* 25 */
    TXOP_concat, /* 26 */
    TXOP_and, /* 27 */
    TXOP_dand, /* 28 */
    TXOP_or, /* 29 */
    TXOP_dor, /* 30 */
    TXOP_not, /* 31 */
    TXOP_plus, /* 32 */
    TXOP_minus, /* 33 */
    TXOP_eq, /* 34 */
    TXOP_ne, /* 35 */
    TXOP_lt, /* 36 */
    TXOP_le, /* 37 */
    TXOP_gt, /* 38 */
    TXOP_ge, /* 39 */
    TXOP_macrocall, /* 40 */
    TXOP_macro_begin, /* 41 */
    TXOP_macro_end, /* 42 */
    TXOP_macro, /* 43 */
    TXOP_function, /* 44 */
    TXOP_funcall, /* 45 */
    TXOP_methodcall_s, /* 46 */
    TXOP_make_array, /* 47 */
    TXOP_make_hash, /* 48 */
    TXOP_enter, /* 49 */
    TXOP_leave, /* 50 */
    TXOP_goto, /* 51 */
    TXOP_depend, /* 52 */
    TXOP_end, /* 53 */
    TXOP_last
}; /* enum tx_opcode_t */

static const tx_exec_t tx_opcode[] = {
    TXCODE_noop,
    TXCODE_move_to_sb,
    TXCODE_move_from_sb,
    TXCODE_save_to_lvar,
    TXCODE_load_lvar_to_sb,
    TXCODE_local_s,
    TXCODE_push,
    TXCODE_pushmark,
    TXCODE_nil,
    TXCODE_literal,
    TXCODE_literal_i,
    TXCODE_fetch_s,
    TXCODE_fetch_lvar,
    TXCODE_fetch_field,
    TXCODE_fetch_field_s,
    TXCODE_print,
    TXCODE_print_raw,
    TXCODE_print_raw_s,
    TXCODE_include,
    TXCODE_for_start,
    TXCODE_for_iter,
    TXCODE_add,
    TXCODE_sub,
    TXCODE_mul,
    TXCODE_div,
    TXCODE_mod,
    TXCODE_concat,
    TXCODE_and,
    TXCODE_dand,
    TXCODE_or,
    TXCODE_dor,
    TXCODE_not,
    TXCODE_plus,
    TXCODE_minus,
    TXCODE_eq,
    TXCODE_ne,
    TXCODE_lt,
    TXCODE_le,
    TXCODE_gt,
    TXCODE_ge,
    TXCODE_macrocall,
    TXCODE_macro_begin,
    TXCODE_macro_end,
    TXCODE_macro,
    TXCODE_function,
    TXCODE_funcall,
    TXCODE_methodcall_s,
    TXCODE_make_array,
    TXCODE_make_hash,
    TXCODE_enter,
    TXCODE_leave,
    TXCODE_goto,
    TXCODE_depend,
    TXCODE_end,
    NULL
}; /* tx_opcode[] */

static const U8 tx_oparg[] = {
    0U, /* noop */
    0U, /* move_to_sb */
    0U, /* move_from_sb */
    TXCODE_W_VAR, /* save_to_lvar */
    TXCODE_W_VAR, /* load_lvar_to_sb */
    TXCODE_W_KEY, /* local_s */
    0U, /* push */
    0U, /* pushmark */
    0U, /* nil */
    TXCODE_W_SV, /* literal */
    TXCODE_W_INT, /* literal_i */
    TXCODE_W_KEY, /* fetch_s */
    TXCODE_W_VAR, /* fetch_lvar */
    0U, /* fetch_field */
    TXCODE_W_KEY, /* fetch_field_s */
    0U, /* print */
    0U, /* print_raw */
    TXCODE_W_SV, /* print_raw_s */
    0U, /* include */
    TXCODE_W_VAR, /* for_start */
    TXCODE_GOTO, /* for_iter */
    0U, /* add */
    0U, /* sub */
    0U, /* mul */
    0U, /* div */
    0U, /* mod */
    TXCODE_W_SV, /* concat */
    TXCODE_GOTO, /* and */
    TXCODE_GOTO, /* dand */
    TXCODE_GOTO, /* or */
    TXCODE_GOTO, /* dor */
    0U, /* not */
    0U, /* plus */
    0U, /* minus */
    0U, /* eq */
    0U, /* ne */
    0U, /* lt */
    0U, /* le */
    0U, /* gt */
    0U, /* ge */
    0U, /* macrocall */
    TXCODE_W_KEY, /* macro_begin */
    0U, /* macro_end */
    TXCODE_W_KEY, /* macro */
    TXCODE_W_KEY, /* function */
    0U, /* funcall */
    TXCODE_W_KEY, /* methodcall_s */
    0U, /* make_array */
    0U, /* make_hash */
    0U, /* enter */
    0U, /* leave */
    TXCODE_GOTO, /* goto */
    TXCODE_W_SV, /* depend */
    0U, /* end */
}; /* tx_oparg[] */

static void
tx_init_ops(pTHX_ HV* const ops) {
    (void)hv_stores(ops, STRINGIFY(noop), newSViv(TXOP_noop));
    (void)hv_stores(ops, STRINGIFY(move_to_sb), newSViv(TXOP_move_to_sb));
    (void)hv_stores(ops, STRINGIFY(move_from_sb), newSViv(TXOP_move_from_sb));
    (void)hv_stores(ops, STRINGIFY(save_to_lvar), newSViv(TXOP_save_to_lvar));
    (void)hv_stores(ops, STRINGIFY(load_lvar_to_sb), newSViv(TXOP_load_lvar_to_sb));
    (void)hv_stores(ops, STRINGIFY(local_s), newSViv(TXOP_local_s));
    (void)hv_stores(ops, STRINGIFY(push), newSViv(TXOP_push));
    (void)hv_stores(ops, STRINGIFY(pushmark), newSViv(TXOP_pushmark));
    (void)hv_stores(ops, STRINGIFY(nil), newSViv(TXOP_nil));
    (void)hv_stores(ops, STRINGIFY(literal), newSViv(TXOP_literal));
    (void)hv_stores(ops, STRINGIFY(literal_i), newSViv(TXOP_literal_i));
    (void)hv_stores(ops, STRINGIFY(fetch_s), newSViv(TXOP_fetch_s));
    (void)hv_stores(ops, STRINGIFY(fetch_lvar), newSViv(TXOP_fetch_lvar));
    (void)hv_stores(ops, STRINGIFY(fetch_field), newSViv(TXOP_fetch_field));
    (void)hv_stores(ops, STRINGIFY(fetch_field_s), newSViv(TXOP_fetch_field_s));
    (void)hv_stores(ops, STRINGIFY(print), newSViv(TXOP_print));
    (void)hv_stores(ops, STRINGIFY(print_raw), newSViv(TXOP_print_raw));
    (void)hv_stores(ops, STRINGIFY(print_raw_s), newSViv(TXOP_print_raw_s));
    (void)hv_stores(ops, STRINGIFY(include), newSViv(TXOP_include));
    (void)hv_stores(ops, STRINGIFY(for_start), newSViv(TXOP_for_start));
    (void)hv_stores(ops, STRINGIFY(for_iter), newSViv(TXOP_for_iter));
    (void)hv_stores(ops, STRINGIFY(add), newSViv(TXOP_add));
    (void)hv_stores(ops, STRINGIFY(sub), newSViv(TXOP_sub));
    (void)hv_stores(ops, STRINGIFY(mul), newSViv(TXOP_mul));
    (void)hv_stores(ops, STRINGIFY(div), newSViv(TXOP_div));
    (void)hv_stores(ops, STRINGIFY(mod), newSViv(TXOP_mod));
    (void)hv_stores(ops, STRINGIFY(concat), newSViv(TXOP_concat));
    (void)hv_stores(ops, STRINGIFY(and), newSViv(TXOP_and));
    (void)hv_stores(ops, STRINGIFY(dand), newSViv(TXOP_dand));
    (void)hv_stores(ops, STRINGIFY(or), newSViv(TXOP_or));
    (void)hv_stores(ops, STRINGIFY(dor), newSViv(TXOP_dor));
    (void)hv_stores(ops, STRINGIFY(not), newSViv(TXOP_not));
    (void)hv_stores(ops, STRINGIFY(plus), newSViv(TXOP_plus));
    (void)hv_stores(ops, STRINGIFY(minus), newSViv(TXOP_minus));
    (void)hv_stores(ops, STRINGIFY(eq), newSViv(TXOP_eq));
    (void)hv_stores(ops, STRINGIFY(ne), newSViv(TXOP_ne));
    (void)hv_stores(ops, STRINGIFY(lt), newSViv(TXOP_lt));
    (void)hv_stores(ops, STRINGIFY(le), newSViv(TXOP_le));
    (void)hv_stores(ops, STRINGIFY(gt), newSViv(TXOP_gt));
    (void)hv_stores(ops, STRINGIFY(ge), newSViv(TXOP_ge));
    (void)hv_stores(ops, STRINGIFY(macrocall), newSViv(TXOP_macrocall));
    (void)hv_stores(ops, STRINGIFY(macro_begin), newSViv(TXOP_macro_begin));
    (void)hv_stores(ops, STRINGIFY(macro_end), newSViv(TXOP_macro_end));
    (void)hv_stores(ops, STRINGIFY(macro), newSViv(TXOP_macro));
    (void)hv_stores(ops, STRINGIFY(function), newSViv(TXOP_function));
    (void)hv_stores(ops, STRINGIFY(funcall), newSViv(TXOP_funcall));
    (void)hv_stores(ops, STRINGIFY(methodcall_s), newSViv(TXOP_methodcall_s));
    (void)hv_stores(ops, STRINGIFY(make_array), newSViv(TXOP_make_array));
    (void)hv_stores(ops, STRINGIFY(make_hash), newSViv(TXOP_make_hash));
    (void)hv_stores(ops, STRINGIFY(enter), newSViv(TXOP_enter));
    (void)hv_stores(ops, STRINGIFY(leave), newSViv(TXOP_leave));
    (void)hv_stores(ops, STRINGIFY(goto), newSViv(TXOP_goto));
    (void)hv_stores(ops, STRINGIFY(depend), newSViv(TXOP_depend));
    (void)hv_stores(ops, STRINGIFY(end), newSViv(TXOP_end));
} /* tx_register_ops() */
